sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/data/OpenBisConnector.java

```diff
@@ -0,0 +1,112 @@
+
+/**
+ * <b>short description</b>
+ *
+ * <p>detailed description</p>
+ *
+ * @since <version tag>
+ */
+public class OpenBisConnector implements ProjectRepository, SampleRepository {
+
+  private final String sessionToken;
+
+  private final IApplicationServerApi api;
+
+  private static final int TIMEOUT = 10_000;
+
+  private final List<ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project> cachedProjects = new ArrayList<>();
+  private final Map<String, List<ch.ethz.sis.openbis.generic.asapi.v3.dto.sample.Sample>> cachedSamples = new HashMap<>();
+
+
+
+  public OpenBisConnector(Credentials credentials, PortalUser portalUser,
+      String openBisUrl) {
+    this.api = HttpInvokerUtils.createServiceStub(IApplicationServerApi.class,
+        openBisUrl + "/rmi-application-server-v3", TIMEOUT);
+    this.sessionToken = api.loginAs(credentials.getUser(), credentials.getPassword(),
+        portalUser.getAuthProviderId());
+  }
+
+  @Override
+  public List<Project> findAllProjects() {
+    if (cachedProjects.isEmpty()) {
+      ProjectFetchOptions fetchOptions = new ProjectFetchOptions();
+      fetchOptions.withSpace();
+      SearchResult<ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project> projectSearchResult =
+          api.searchProjects(sessionToken, new ProjectSearchCriteria(), fetchOptions);
+
+      List<ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project> projects = projectSearchResult.getObjects().stream()
+          .filter(hasValidProjectCode()).sorted(Comparator.comparing(
+              ch.ethz.sis.openbis.generic.asapi.v3.dto.project.Project::getRegistrationDate).reversed())
+          .collect(Collectors.toList());
+      cachedProjects.addAll(projects);
+    }
+    return cachedProjects.stream()
+        .map(it -> new Project(it.getCode(),
+            Optional.ofNullable(it.getDescription()).orElse("")))
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public List<Sample> findAllSamplesForProject(String projectCode) {
+
+    List<ch.ethz.sis.openbis.generic.asapi.v3.dto.sample.Sample> samples;
+    if (cachedSamples.containsKey(projectCode)) {
+      samples = cachedSamples.get(projectCode);
+    } else {
+      SampleSearchCriteria sampleSearchCriteria = new SampleSearchCriteria();
+      sampleSearchCriteria.withCode().thatStartsWith(projectCode);
+      sampleSearchCriteria.withType().withCode().thatEquals("Q_TEST_SAMPLE");
+      SampleFetchOptions fetchOptions = new SampleFetchOptions();
+      fetchOptions.withType();
+      fetchOptions.withProperties();
+      SearchResult<ch.ethz.sis.openbis.generic.asapi.v3.dto.sample.Sample> sampleSearchResult = api.searchSamples(
+          sessionToken, sampleSearchCriteria, fetchOptions);
+      samples = sampleSearchResult.getObjects();
+      cachedSamples.put(projectCode, samples);
+    }
+    return samples.stream()
+        .filter(hasValidSampleCode())
+        .filter(notAnEntity())
```

Comment:
NIT: the regex already excludes entitites

Reply:
I think it would be better to have it explicitly here. Otherwise it might be overlooked in the future. Filtering does not lead to overhead here as java streams lazyly evaluates.